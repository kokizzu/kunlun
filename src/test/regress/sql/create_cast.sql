--
-- CREATE_CAST
--

-- Create some types to test with
--DDL_STATEMENT_BEGIN--
CREATE TYPE casttesttype;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
CREATE FUNCTION casttesttype_in(cstring)
   RETURNS casttesttype
   AS 'textin'
   LANGUAGE internal STRICT IMMUTABLE;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
CREATE FUNCTION casttesttype_out(casttesttype)
   RETURNS cstring
   AS 'textout'
   LANGUAGE internal STRICT IMMUTABLE;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
CREATE TYPE casttesttype (
   internallength = variable,
   input = casttesttype_in,
   output = casttesttype_out,
   alignment = int4
);
--DDL_STATEMENT_END--

-- a dummy function to test with
--DDL_STATEMENT_BEGIN--
CREATE FUNCTION casttestfunc(casttesttype) RETURNS int4 LANGUAGE SQL AS
$$ SELECT 1; $$;
--DDL_STATEMENT_END--

SELECT casttestfunc('foo'::text); -- fails, as there's no cast

-- Try binary coercion cast
--DDL_STATEMENT_BEGIN--
CREATE CAST (text AS casttesttype) WITHOUT FUNCTION;
--DDL_STATEMENT_END--
SELECT casttestfunc('foo'::text); -- doesn't work, as the cast is explicit
SELECT casttestfunc('foo'::text::casttesttype); -- should work
--DDL_STATEMENT_BEGIN--
DROP CAST (text AS casttesttype); -- cleanup
--DDL_STATEMENT_END--

-- Try IMPLICIT binary coercion cast
--DDL_STATEMENT_BEGIN--
CREATE CAST (text AS casttesttype) WITHOUT FUNCTION AS IMPLICIT;
--DDL_STATEMENT_END--
SELECT casttestfunc('foo'::text); -- Should work now

-- Try I/O conversion cast.
SELECT 1234::int4::casttesttype; -- No cast yet, should fail
--DDL_STATEMENT_BEGIN--

CREATE CAST (int4 AS casttesttype) WITH INOUT;
--DDL_STATEMENT_END--
SELECT 1234::int4::casttesttype; -- Should work now
--DDL_STATEMENT_BEGIN--

DROP CAST (int4 AS casttesttype);
--DDL_STATEMENT_END--

-- Try cast with a function

--DDL_STATEMENT_BEGIN--
CREATE FUNCTION int4_casttesttype(int4) RETURNS casttesttype LANGUAGE SQL AS
$$ SELECT ('foo'::text || $1::text)::casttesttype; $$;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
CREATE CAST (int4 AS casttesttype) WITH FUNCTION int4_casttesttype(int4) AS IMPLICIT;
--DDL_STATEMENT_END--
SELECT 1234::int4::casttesttype; -- Should work now
