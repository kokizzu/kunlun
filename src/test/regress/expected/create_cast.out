--
-- CREATE_CAST
--
-- Create some types to test with
CREATE TYPE casttesttype;
CREATE TYPE
CREATE FUNCTION casttesttype_in(cstring)
   RETURNS casttesttype
   AS 'textin'
   LANGUAGE internal STRICT IMMUTABLE;
psql:sql/create_cast.sql:11: NOTICE:  return type casttesttype is only a shell
CREATE FUNCTION
CREATE FUNCTION casttesttype_out(casttesttype)
   RETURNS cstring
   AS 'textout'
   LANGUAGE internal STRICT IMMUTABLE;
psql:sql/create_cast.sql:15: NOTICE:  argument type casttesttype is only a shell
CREATE FUNCTION
CREATE TYPE casttesttype (
   internallength = variable,
   input = casttesttype_in,
   output = casttesttype_out,
   alignment = int4
);
CREATE TYPE
-- a dummy function to test with
CREATE FUNCTION casttestfunc(casttesttype) RETURNS int4 LANGUAGE SQL AS
$$ SELECT 1; $$;
CREATE FUNCTION
SELECT casttestfunc('foo'::text); -- fails, as there's no cast
 casttestfunc 
--------------
            1
(1 row)

-- Try binary coercion cast
CREATE CAST (text AS casttesttype) WITHOUT FUNCTION;
CREATE CAST
SELECT casttestfunc('foo'::text); -- doesn't work, as the cast is explicit
 casttestfunc 
--------------
            1
(1 row)

SELECT casttestfunc('foo'::text::casttesttype); -- should work
 casttestfunc 
--------------
            1
(1 row)

DROP CAST (text AS casttesttype); -- cleanup
DROP CAST
-- Try IMPLICIT binary coercion cast
CREATE CAST (text AS casttesttype) WITHOUT FUNCTION AS IMPLICIT;
CREATE CAST
SELECT casttestfunc('foo'::text); -- Should work now
 casttestfunc 
--------------
            1
(1 row)

-- Try I/O conversion cast.
SELECT 1234::int4::casttesttype; -- No cast yet, should fail
psql:sql/create_cast.sql:41: ERROR:  cannot cast type integer to casttesttype
LINE 1: SELECT 1234::int4::casttesttype;
                         ^
CREATE CAST (int4 AS casttesttype) WITH INOUT;
CREATE CAST
SELECT 1234::int4::casttesttype; -- Should work now
 casttesttype 
--------------
 1234
(1 row)

DROP CAST (int4 AS casttesttype);
DROP CAST
-- Try cast with a function
CREATE FUNCTION int4_casttesttype(int4) RETURNS casttesttype LANGUAGE SQL AS
$$ SELECT ('foo'::text || $1::text)::casttesttype; $$;
CREATE FUNCTION
CREATE CAST (int4 AS casttesttype) WITH FUNCTION int4_casttesttype(int4) AS IMPLICIT;
CREATE CAST
SELECT 1234::int4::casttesttype; -- Should work now
 casttesttype 
--------------
 foo1234
(1 row)

